src/zarr/hierarchy.py

```diff
@@ -0,0 +1,366 @@
+    elif hasattr(data, "data_type") and not hasattr(data, "dtype"):
+        return data.data_type
+    else:
+        msg = (
+            f"Could not infer the data_type attribute from {data}. "
+            "Expected either an object with a `dtype` attribute, "
+            "or an object with a `data_type` attribute."
+        )
+        raise ValueError(msg)
+
+
+def auto_attributes(data: Any) -> Any:
+    """
+    Guess attributes from:
+        input with an `attrs` attribute, or
+        input with an `attributes` attribute,
+        or anything (returning {})
+    """
+    if hasattr(data, "attrs"):
+        return data.attrs
+    if hasattr(data, "attributes"):
+        return data.attributes
+    return {}
+
+
+def auto_chunk_key_encoding(data: Any) -> Any:
+    if hasattr(data, "chunk_key_encoding"):
+        return data.chunk_key_encoding
+    return DefaultChunkKeyEncoding()
+
+
+def auto_fill_value(data: Any) -> Any:
+    """
+    Guess fill value from an input with a `fill_value` attribute, returning 0 otherwise.
+    """
+    if hasattr(data, "fill_value"):
+        return data.fill_value
+    return 0
+
+
+def auto_codecs(data: Any) -> Any:
+    """
+    Guess compressor from an input with a `compressor` attribute, returning `None` otherwise.
+    """
+    if hasattr(data, "codecs"):
+        return data.codecs
+    return (BytesCodec(),)
+
+
+def auto_dimension_names(data: Any) -> Any:
+    """
+    If the input has a `dimension_names` attribute, return it, otherwise
+    return None.
+    """
+
+    if hasattr(data, "dimension_names"):
+        return data.dimension_names
+    return None
+
+
+def auto_chunk_grid(data: Any) -> Any:
+    """
+    Guess a chunk grid from:
+      input with a `chunk_grid` attribute,
+      input with a `chunksize` attribute, or
+      input with a `chunks` attribute, or,
+      input with `shape` and `dtype` attributes
+    """
+    if hasattr(data, "chunk_grid"):
+        # more a statement of intent than anything else
+        return data.chunk_grid
+    if hasattr(data, "chunksize"):
+        chunks = data.chunksize
+    elif hasattr(data, "chunks"):
+        chunks = data.chunks
+    else:
+        chunks = guess_chunks(data.shape, np.dtype(data.dtype).itemsize)
+    return RegularChunkGrid(chunk_shape=chunks)
+
+
+class ArrayModel(ArrayV3Metadata):
+    """
+    A model of a Zarr v3 array.
+    """
+
+    @classmethod
+    def from_stored(cls: type[Self], node: Array) -> Self:
+        """
+        Create an array model from a stored array.
+        """
+        return cls.from_dict(node.metadata.to_dict())
+
+    def to_stored(self, store_path: StorePath, exists_ok: bool = False) -> Array:
+        """
+        Create a stored version of this array.
+        """
+        # exists_ok kwarg is unhandled until we wire it up to the
+        # array creation routines
+
+        return Array.from_dict(store_path=store_path, data=self.to_dict())
+
+    @classmethod
+    def from_array(
```

Comment:
this is a convenience method that makes it less painful to create `ArrayModel` instances, because it uses as many defaults / inferred values as possible.